akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedBehaviorRecoveryTimeoutSpec.scala

```diff
@@ -0,0 +1,108 @@
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
+import akka.persistence.RecoveryTimedOut
+import akka.persistence.journal.SteppingInmemJournal
+import akka.persistence.typed.PersistenceId
+import akka.persistence.typed.RecoveryFailed
+import akka.persistence.typed.internal.JournalFailureException
+import akka.testkit.EventFilter
+import akka.testkit.TestEvent.Mute
+import com.typesafe.config.Config
+import com.typesafe.config.ConfigFactory
+import org.scalatest.WordSpecLike
+
+object EventSourcedBehaviorRecoveryTimeoutSpec {
+
+  val journalId = "event-sourced-behavior-recovery-timeout-spec"
+
+  def config: Config =
+    SteppingInmemJournal
+      .config(journalId)
+      .withFallback(ConfigFactory.parseString("""
+        akka.persistence.journal.stepping-inmem.recovery-event-timeout=1s
+        """))
+      .withFallback(ConfigFactory.parseString(s"""
+        akka.loglevel = INFO
+        akka.loggers = [akka.testkit.TestEventListener]
+        """))
+
+  def testBehavior(persistenceId: PersistenceId, probe: ActorRef[AnyRef]): Behavior[String] =
+    Behaviors.setup { _ =>
+      EventSourcedBehavior[String, String, String](
+        persistenceId,
+        emptyState = "",
+        commandHandler = (_, command) => Effect.persist(command).thenRun(_ => probe ! command),
+        eventHandler = (state, evt) => state + evt).receiveSignal {
+        case RecoveryFailed(cause) =>
+          probe ! cause
+      }
+    }
+
+}
+
+class EventSourcedBehaviorRecoveryTimeoutSpec
+    extends ScalaTestWithActorTestKit(EventSourcedBehaviorRecoveryTimeoutSpec.config)
+    with WordSpecLike {
+
+  import EventSourcedBehaviorRecoveryTimeoutSpec._
+
+  val pidCounter = new AtomicInteger(0)
+  private def nextPid(): PersistenceId = PersistenceId(s"c${pidCounter.incrementAndGet()})")
+
+  import akka.actor.typed.scaladsl.adapter._
+  // needed for SteppingInmemJournal.step
+  private implicit val untypedSystem: akka.actor.ActorSystem = system.toUntyped
+
+  untypedSystem.eventStream.publish(Mute(EventFilter.warning(start = "No default snapshot store", occurrences = 1)))
+
+  "The recovery timeout" must {
+
+    "fail recovery if timeout is not met when recovering" in {
+      val probe = createTestProbe[AnyRef]()
+      val pid = nextPid()
+      val persisting = spawn(testBehavior(pid, probe.ref))
+
+      probe.awaitAssert(SteppingInmemJournal.getRef(journalId), 3.seconds)
+      val journal = SteppingInmemJournal.getRef(journalId)
+
+      // initial read highest
+      SteppingInmemJournal.step(journal)
+
+      persisting ! "A"
+      SteppingInmemJournal.step(journal)
+      probe.expectMessage("A")
+
+      testKit.stop(persisting)
+      probe.expectTerminated(persisting)
+
+      // now replay, but don't give the journal any tokens to replay events
+      // so that we cause the timeout to trigger
+      EventFilter[JournalFailureException](start = "Exception during recovery.", occurrences = 1).intercept {
```

Comment:
More specific test error message:

Reply:
Difficult to grab something more specific with the EventFilter here. The full logging is:
```
[ERROR] [03/26/2019 16:14:41.691] [EventSourcedBehaviorRecoveryTimeoutSpec-akka.actor.default-dispatcher-2] [akka://EventSourcedBehaviorRecoveryTimeoutSpec/user/$b] Supervisor StopSupervisor saw failure: Exception during recovery. Last known sequence number [0]. PersistenceId [c1)]
akka.persistence.typed.internal.JournalFailureException: Exception during recovery. Last known sequence number [0]. PersistenceId [c1)]
	at akka.persistence.typed.internal.ReplayingEvents.onRecoveryFailure(ReplayingEvents.scala:183)
	at akka.persistence.typed.internal.ReplayingEvents.onRecoveryTick(ReplayingEvents.scala:142)
	at akka.persistence.typed.internal.ReplayingEvents.onMessage(ReplayingEvents.scala:77)
	at akka.persistence.typed.internal.ReplayingEvents.onMessage(ReplayingEvents.scala:59)
	at akka.actor.typed.scaladsl.AbstractBehavior.receive(AbstractBehavior.scala:59)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:429)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:399)
	at akka.actor.typed.internal.InterceptorImpl$$anon$2.apply(InterceptorImpl.scala:53)
	at akka.actor.typed.internal.WidenedInterceptor.aroundReceive(InterceptorImpl.scala:208)
	at akka.actor.typed.internal.InterceptorImpl.receive(InterceptorImpl.scala:80)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:429)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:399)
	at akka.actor.typed.internal.InterceptorImpl$$anon$2.apply(InterceptorImpl.scala:53)
	at akka.persistence.typed.internal.EventSourcedBehaviorImpl$$anon$1.aroundReceive(EventSourcedBehaviorImpl.scala:136)
	at akka.actor.typed.internal.InterceptorImpl.receive(InterceptorImpl.scala:80)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:429)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:399)
	at akka.actor.typed.internal.InterceptorImpl$$anon$2.apply(InterceptorImpl.scala:53)
	at akka.actor.typed.internal.SimpleSupervisor.aroundReceive(Supervision.scala:103)
	at akka.actor.typed.internal.InterceptorImpl.receive(InterceptorImpl.scala:80)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:429)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:399)
	at akka.actor.typed.internal.adapter.ActorAdapter.akka$actor$typed$internal$adapter$ActorAdapter$$handleMessage(ActorAdapter.scala:91)
	at akka.actor.typed.internal.adapter.ActorAdapter$$anonfun$running$1.applyOrElse(ActorAdapter.scala:86)
	at akka.actor.Actor.aroundReceive(Actor.scala:539)
	at akka.actor.Actor.aroundReceive$(Actor.scala:537)
	at akka.actor.typed.internal.adapter.ActorAdapter.aroundReceive(ActorAdapter.scala:41)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:610)
	at akka.actor.ActorCell.invoke(ActorCell.scala:579)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:268)
	at akka.dispatch.Mailbox.run(Mailbox.scala:229)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:241)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: akka.persistence.RecoveryTimedOut: Replay timed out, didn't get event within [1000 milliseconds], highest sequence number seen [0]
```

but that "Replay timed out" is only part of the cause message.

I have included the cause message in the logged message for recovery failures, better with too much then too little. See a181140